ledger/complete/wal/cached_compactor.go

```diff
@@ -0,0 +1,435 @@
+	forest           *mtrie.Forest
+	forestSegmentNum int
+
+	segmentUpdateCh <-chan *segmentTries
+	stopCh          chan struct{}
+	checkpointCh    chan<- int
+}
+
+func newSegmentsTrieCompactor(
+	logger zerolog.Logger,
+	checkpointer *Checkpointer,
+	segmentUpdateCh <-chan *segmentTries,
+	checkpointCh chan<- int,
+	tries []*trie.MTrie,
+	segmentNum int,
+	checkpointForestCapacity int,
+	checkpointDistance uint,
+) (*segmentsTrieCompactor, error) {
+
+	forest, err := mtrie.NewForest(checkpointForestCapacity, &metrics.NoopCollector{}, nil)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create Forest: %w", err)
+	}
+	err = forest.AddTries(tries)
+	if err != nil {
+		return nil, fmt.Errorf("cannot add tries to forest: %w", err)
+	}
+
+	lastCheckpointNum, err := checkpointer.LatestCheckpoint()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get last checkpointed number: %w", err)
+	}
+
+	return &segmentsTrieCompactor{
+		checkpointer:       checkpointer,
+		logger:             logger,
+		checkpointDistance: checkpointDistance,
+		lastCheckpointNum:  lastCheckpointNum,
+		forest:             forest,
+		forestSegmentNum:   segmentNum,
+		segmentUpdateCh:    segmentUpdateCh,
+		checkpointCh:       checkpointCh,
+		stopCh:             make(chan struct{}),
+	}, nil
+}
+
+type checkpointResult struct {
+	checkpointNum int
+	err           error
+}
+
+func (c *segmentsTrieCompactor) stop() {
+	c.stopCh <- struct{}{}
+}
+
+func (c *segmentsTrieCompactor) start() {
+
+	checkpointSem := semaphore.NewWeighted(1) // limit to 1 checkpointing goroutine
+	checkpointResultChan := make(chan checkpointResult)
+
+Loop:
+	for {
+		select {
+
+		case <-c.stopCh:
+			break Loop
+
+		case tries := <-c.segmentUpdateCh:
+			triesToBeCheckpointed, checkpointNum, err := c.update(tries.tries, tries.segmentNum)
+			if err != nil {
+				c.logger.Error().Err(err).Msg("error updating cache")
+				continue
+			}
+
+			if len(triesToBeCheckpointed) > 0 {
+				if checkpointSem.TryAcquire(1) {
+					go func() {
+						defer checkpointSem.Release(1)
+						err := createCheckpoint(c.checkpointer, c.logger, triesToBeCheckpointed, checkpointNum)
+						checkpointResultChan <- checkpointResult{checkpointNum, err}
+					}()
+				}
+			}
+
+		case cpResult := <-checkpointResultChan:
+			if cpResult.err != nil {
+				c.logger.Error().Err(cpResult.err).Msg("error checkpointing")
+				continue
+			}
+
+			c.Lock()
+			c.lastCheckpointNum = cpResult.checkpointNum
+			c.Unlock()
+
+			c.checkpointCh <- cpResult.checkpointNum
+		}
+	}
+
+	// Drain remaining segment updates to remove trie references.
+	for range c.segmentUpdateCh {
+	}
+}
+
+// update adds tries in given segment to forest, and returns tries to be checkpointed
+// with checkpoint number if enough segments are finalized.
+func (c *segmentsTrieCompactor) update(tries []*trie.MTrie, segmentNum int) ([]*trie.MTrie, int, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	err := c.forest.AddTries(tries)
```

Comment:
Adding all the trie nodes of a segment to the forest, and check whether checkpointing should be triggered by comparing with the checkpoint distance config. 

If checkpoint should be triggered, then it should return the last X trie nodes to be included in a new checkpoint file.

Reply:
Yes, exactly!  :+1:

Reply:
At this point, we assume that the tries are consistent with those in the WAL files. 

Consistent means both the total number and the order are consistent. 

I wonder if possible that we can sanity check that with low cost.

How is the WAL file encoded? Is it possible to scan through a range of WAL files and only read the file header (or footer) about the number of trie nodes, and their root hash without loading the entire file into memory.

Reply:
> How is the WAL file encoded? Is it possible to scan through a range of WAL files and only read the file header (or footer) about the number of trie nodes, and their root hash without loading the entire file into memory?

I think WAL file uses proprietary encoding for records.  I don't think there is a quick way to scan information without reading the entire file.  

> At this point, we assume that the tries are consistent with those in the WAL files.
> I wonder if possible that we can sanity check that assumption with low cost.

I think PR #2792 enforce this assumption with communication between `Compactor` and `Ledger.Set`.

Also `TestCompactorAccuracy` in PR #2792 passes.  It creates ~30 segments and triggers checkpointing every 5 segments.  The tests expects checkpointed tries to match replayed tries.  Replayed tries are tries updated by replaying all WAL segments (from segment 0, ignoring prior checkpoints) to the checkpoint number.  This verifies that checkpointed tries are snapshot of segments and at segment boundary.

`TestCompactorConcurrency` also tests `Set` in parallel.